ocrd_utils/ocrd_utils/logging.py

```diff
@@ -71,63 +74,84 @@ def setOverrideLogLevel(lvl, silent=False):
     lvl = getLevelName(lvl)
     global _overrideLogLevel # pylint: disable=global-statement
     _overrideLogLevel = lvl
-    logging.getLogger('').setLevel(lvl)
     for loggerName in logging.Logger.manager.loggerDict:
         logger = logging.Logger.manager.loggerDict[loggerName]
         if isinstance(logger, logging.PlaceHolder):
             continue
         logger.setLevel(logging.NOTSET)
+    root_logger.setLevel(lvl)
 
 def getLogger(*args, **kwargs):
     """
     Wrapper around ``logging.getLogger`` that respects `overrideLogLevel <#setOverrideLogLevel>`_.
     """
+    if not _initialized_flag:
+        initLogging()
+        logging.getLogger('').critical('getLogger was called before initLogging. Source of the call:')
+        for line in [x for x in format_stack(limit=2)[0].split('\n') if x]:
+            logging.getLogger('').critical(line)
+    name = args[0]
     logger = logging.getLogger(*args, **kwargs)
-    if _overrideLogLevel is not None:
+    if _overrideLogLevel and name:
         logger.setLevel(logging.NOTSET)
     return logger
 
-# Default logging config
-
 def initLogging():
     """
     Reset root logger, read logging configuration if exists, otherwise use basicConfig
     """
+    global _initialized_flag # pylint: disable=global-statement
+    if _initialized_flag:
+        logging.getLogger('').critical('initLogging was called multiple times. Source of latest call:')
+        for line in [x for x in format_stack(limit=2)[0].split('\n') if x]:
+            logging.getLogger('').critical(line)
 
+    logging.disable(logging.NOTSET)
     for handler in logging.root.handlers[:]:
         logging.root.removeHandler(handler)
 
-    global _overrideLogLevel # pylint: disable=global-statement
-    _overrideLogLevel = None
-
     CONFIG_PATHS = [
         os.path.curdir,
         os.path.join(os.path.expanduser('~')),
         '/etc',
     ]
-    for p in CONFIG_PATHS:
-        config_file = os.path.join(p, 'ocrd_logging.conf')
-        if os.path.exists(config_file):
-            logging.config.fileConfig(config_file)
-            return
-
-    logging.basicConfig(
-        level=logging.INFO,
-        format=LOG_FORMAT,
-        datefmt=LOG_TIMEFMT)
-    logging.getLogger('').setLevel(logging.INFO)
-    #  logging.getLogger('ocrd.resolver').setLevel(logging.INFO)
-    #  logging.getLogger('ocrd.resolver.download_to_directory').setLevel(logging.INFO)
-    #  logging.getLogger('ocrd.resolver.add_files_to_mets').setLevel(logging.INFO)
-    logging.getLogger('PIL').setLevel(logging.INFO)
-    # To cut back on the `Self-intersection at or near point` INFO messages
-    logging.getLogger('shapely.geos').setLevel(logging.ERROR)
-
+    config_file = next((f for f \
+            in [os.path.join(p, 'ocrd_logging.conf') for p in CONFIG_PATHS] \
+            if os.path.exists(f)),
+            None)
+    if config_file:
+        logging.info("Picked up logging config at %s" % config_file)
```

Comment:
I think we discussed that line before. The problem is that it will always show a message – even if the config says not to. Think about bashlib's `ocrd bashlib log debug` – that would now create 1 info message unconditionally plus debug messages if the level permits it.

So I recommend any combination of the following counter-measures:
1. move that 1 line below
1. downgrade to `debug`
1. use a non-root logger like `logging.getLogger('ocrd.logging')`

Reply:
> that would now create 1 info message unconditionally

Good point. 815fc076